File: src/courses/data/catalog/02-dsa/dsa-core/implementations/csharp/bst.cs
/**
 * C# Implementation: Binary Search Tree (BST)
 * Invariant: Balanced O(log N) Search/Insert, Degenerate O(N)
 */
using System;
using System.Collections.Generic;

namespace Seed.DSA.Core
{
    public class BinarySearchTreeBST<T>
    {
        private readonly List<T> _items = new List<T>();

        // TODO: Implement algorithm operations
        public void Add(T item)
        {
            _items.Add(item);
        }

        public int Count => _items.Count;
    }
}

File: src/courses/data/catalog/02-dsa/dsa-core/implementations/python/bst.py
"""
Python Implementation: Binary Search Tree (BST)
Invariant: Balanced O(log N) Search/Insert, Degenerate O(N)
"""
from typing import Generic, TypeVar, List, Optional

T = TypeVar('T')

class BinarySearchTreeBST(Generic[T]):
    def __init__(self) -> None:
        self._items: List[T] = []

    # TODO: Implement core operational method
    def insert(self, value: T) -> None:
        self._items.append(value)

    def __len__(self) -> int:
        return len(self._items)
